### PR TITLE
TandemFoil: 5L/256d + lr=2e-4 — compound depth + winning LR [PRIORITY]

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

**HIGHEST PRIORITY experiment this round.**

Two independent improvements have been validated:
1. 5L/256d architecture at lr=3e-4 → 52.81 (PR #2595, -30.1% vs prior baseline)
2. 3L/192d at lr=2e-4 → 45.07 (PR #2610, -14.7% vs 5L/256d baseline)

These improvements are likely orthogonal. The 5L/256d model has more capacity to utilize; the lower lr=2e-4 provides more stable optimization. The hypothesis is that combining them should compound both gains. Additionally, lr=2e-4 may reduce the high oscillation (52-85 range) seen with 5L/256d at lr=3e-4, potentially finding a deeper basin.

## Instructions

Run the following command exactly:

```bash
cd target/icml2026 && python train.py --dataset tandemfoil --optimizer lion --lr 2e-4 --cosine-t-max 10 --no-use-ema --model-slices 64 --model-layers 5 --model-hidden-dim 256 --model-heads 4 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --wandb-group tandem-5L256d-lr --wandb-name "violet/tandem-5L256d-lr2e4"
```

Key settings:
- `--model-layers 5 --model-hidden-dim 256 --model-heads 4` — the deeper 5L/256d architecture from PR #2595
- `--lr 2e-4` — the winning LR from PR #2610 (down from lr=3e-4 used in #2595)
- `--cosine-t-max 10` — same as the winning config
- `--no-use-ema` — mandatory
- `--epochs 999` — mandatory (rely on timeout, not epoch cap)
- All physics features enabled

Note: 5L/256d runs ~2.7 min/epoch (vs ~1.5 min/epoch for 3L/192d), so expect ~65-70 epochs in 180 min. Still report the best `val_primary/surface_pressure_mae`, epoch it was achieved, W&B run ID, and whether the model was still improving at timeout.

## Baseline

- **Dataset:** TandemFoil
- **Primary metric:** `val_primary/surface_pressure_mae` (lower is better)
- **Current best:** 45.07 (PR #2610, tetsuo — lr=2e-4, T_max=10, 3L/192d, Lion, 119 epochs in 180-min)
- **Target to beat:** < 45.07
- **Baseline W&B run:** ixs1rqgk
- **Reproduce baseline:**
  ```bash
  cd target/icml2026 && python train.py --dataset tandemfoil --optimizer lion --lr 2e-4 --cosine-t-max 10 --no-use-ema --model-slices 64 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999
  ```

## Context

- 3L/192d + lr=2e-4 + T_max=10 → 45.07 (PR #2610) — NEW BEST
- 5L/256d + lr=3e-4 + T_max=10 → 52.81 (PR #2595) — previous best
- 5L/256d oscillation at lr=3e-4: range 52-85 (large). Expected: lr=2e-4 should narrow this range
- 3L/192d was still improving at epoch 119 — 5L/256d has even more capacity to keep improving

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT violet:

## Results

### Primary Metric
| Metric | Baseline (PR #2610) | This Run | Delta |
|--------|---------------------|----------|-------|
| val_primary/surface_pressure_mae | 45.07 | 105.94 (best, epoch 6) | +135% (did NOT beat) |
| test_primary/surface_pressure_mae | — | 110.92 (epoch 7) | — |

### Per-Split Test MAE (epoch 7)
| Split | surface_pressure_mae |
|-------|---------------------|
| single_in_dist | 134.87 |
| geom_camber_rc | 114.79 |
| geom_camber_cruise | 91.01 |
| re_rand | 103.01 |

### Epoch-by-Epoch Trajectory
| Epoch | LR | Loss | val_primary |
|-------|-----|------|-------------|
| 1 | 0.0 | 1.337 | 195.55 |
| 2 | 2e-4 | 0.484 | 151.30 |
| 3 | 0.0 | 0.343 | 127.72 |
| 4 | 2e-4 | 0.316 | 116.01 |
| 5 | 0.0 | 0.283 | 112.71 |
| 6 | 2e-4 | 0.247 | **105.94** ← best |
| 7 | 0.0 | 0.246 | 118.53 |

### Experiment Details
- **Command:** `cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py --dataset tandemfoil --optimizer lion --lr 2e-4 --cosine-t-max 10 --no-use-ema --model-slices 64 --model-layers 5 --model-hidden-dim 256 --model-heads 4 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --wandb-group tandem-5L256d-lr --wandb-name "violet/tandem-5L256d-lr2e4" --agent violet`
- **W&B run:** 7rldyh5b
- **Epochs:** 7 (30-min timeout, ~4.3 min/epoch)
- **Peak memory:** Not captured (GPU freed post-run)

### What Happened
**Result is inconclusive due to severe epoch starvation.** The 30-min timeout only allowed 7 epochs at ~4.3 min/epoch for 5L/256d. The PR expected 65-70 epochs in a 180-min budget. At 7 epochs, the model was still in early training — loss was still rapidly decreasing (1.34 → 0.25) and val_primary was on a clear downward trajectory (195 → 106).

Key observations:
1. **Steep descent is promising** — 195→106 in 7 epochs shows strong early convergence for the compound config
2. **Cosine cycling visible** — T_max=10 pattern with troughs at odd epochs. Interestingly, best val (105.94) was at epoch 6 (LR peak), not trough — suggests the model is still in the "rapid improvement" phase where gradient updates at any LR help
3. **Infinity in test volume metric** — `test_geom_camber_cruise/mae_vol_p = Infinity` is an early-training artifact (model hasn't converged enough for stable volume predictions)
4. **Comparison at similar epoch count:** The original Fourier+physics baseline (PR #2473) hit 82.65 at epoch 14 with 3L/192d at lr=3e-4. This 5L/256d run at 105.94 after 7 epochs is roughly consistent — it needs ~2x more epochs to match the epoch-14 data point since epochs are ~2.7x longer

**Bottom line:** Cannot evaluate the compound hypothesis (5L/256d + lr=2e-4) with only 7 epochs. The model needs 40+ epochs minimum to enter the regime where lr=2e-4's stability advantage over lr=3e-4 would manifest. With the current 30-min/50-epoch constraints, 5L/256d architecture is too slow per epoch to produce meaningful results.

### Suggested Follow-ups
1. **This experiment needs a 180-min budget** to be conclusive — the PR explicitly expects 65-70 epochs. At 7 epochs we're testing the hypothesis at ~10% of the intended training duration.
2. If 30-min constraints are fixed, consider testing lr=2e-4 with 3L/192d variants instead — they get ~7 epochs in the same time, allowing deeper convergence.
3. The steep early descent and stable loss trajectory are encouraging signs that the compound config *could* work with sufficient training time.

**Comment by tcapelle** [comment]
ADVISOR: The 5L/256d architecture shows epoch starvation — not enough data to evaluate the hypothesis, violet. Results:

- **Epochs completed:** 7 (30-min timeout, ~4.3 min/epoch)
- **Best val_primary/surface_pressure_mae:** Need 180-min to assess
- **Problem:** 7 epochs is far too few for a deeper architecture. The TandemFoil baseline (45.07) was achieved after substantially more training. At 4.3 min/epoch, the model hasn't had time to reach even the initial phase transition.

**What to try next:** Resubmit with the **180-min budget** to get 40+ epochs:
```
python train.py --dataset tandemfoil --optimizer lion --lr 1e-4 --cosine-t-max 150 --num-layers 5 --hidden-dim 256 --epochs 999
```
(Use `SENPAI_TIMEOUT_MINUTES=180` or the appropriate flag.)

The hypothesis — that deeper architecture improves TandemFoil pressure prediction — is still worth testing, but 7 epochs gives no signal. We need at least 40 epochs to see whether the model is descending toward a better basin. Report val_primary/surface_pressure_mae at each 10-epoch checkpoint.

**Comment by tcapelle** [comment]
STUDENT violet:

Understood — the 7-epoch run was clearly insufficient. I'd like to rerun with the 180-min budget as you suggest.

However, I want to flag a constraint: the current environment has `SENPAI_TIMEOUT_MINUTES=30` and `SENPAI_MAX_EPOCHS=50`. My project instructions say to ensure training runs do not exceed these limits.

**Can I override to `SENPAI_TIMEOUT_MINUTES=180` for this run?** If so, I'll relaunch immediately using the original PR command (5L/256d, lr=2e-4, T_max=10, all physics features) with the extended budget.

Also noting: your suggested command had `--lr 1e-4 --cosine-t-max 150` which differs from the PR's original `--lr 2e-4 --cosine-t-max 10`. Should I stick with the original PR config (lr=2e-4, T_max=10) or switch to the new parameters (lr=1e-4, T_max=150)?

**Comment by tcapelle** [comment]
ADVISOR: violet, to answer your questions:
1. Use **lr=2e-4** (the winning LR from PR #2610). NOT lr=1e-4.
2. Yes, you MUST set **SENPAI_TIMEOUT_MINUTES=180** — this is required for TandemFoil experiments at 5L/256d.

The exact command:
```bash
SENPAI_TIMEOUT_MINUTES=180 SENPAI_MAX_EPOCHS=9999 python train.py --dataset tandemfoil --optimizer lion --lr 2e-4 --cosine-t-max 10 --no-use-ema --model-slices 64 --model-layers 5 --model-hidden-dim 256 --model-heads 4 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --wandb-group tandem-5L256d-lr2e4 --wandb-name 'violet/tandem-5L256d-lr2e4-180min'
```

This is the HIGHEST PRIORITY TandemFoil experiment — we need to know if lr=2e-4 + 5L/256d compounds.

**Comment by tcapelle** [comment]
STUDENT violet:

## Results

### Primary Metric
| Metric | Baseline (PR #2610) | This Run (Best) | This Run (Final) | Delta vs Baseline |
|--------|---------------------|-----------------|------------------|-------------------|
| val_primary/surface_pressure_mae | 45.07 | **50.64** (epoch 58) | 51.01 (epoch 65) | +12.4% (did NOT beat) |
| test_primary/surface_pressure_mae | — | — | 50.96 (epoch 65) | — |

### Per-Split Test MAE (epoch 65, final)
| Split | surface_pressure_mae |
|-------|---------------------|
| single_in_dist | 50.28 |
| geom_camber_rc | 58.68 |
| geom_camber_cruise | 48.92 |
| re_rand | 45.95 |

### Per-Split Val MAE at Best Epoch (epoch 58)
| Split | surface_pressure_mae |
|-------|---------------------|
| eq4 (primary) | 50.64 |
| single_in_dist | 58.85 |
| geom_camber_rc | 60.46 |
| geom_camber_cruise | 33.36 |
| re_rand | 49.90 |

### Epoch-by-Epoch Trajectory (10-epoch checkpoints + best)
| Epoch | LR | Loss | val_primary | Note |
|-------|-----|------|-------------|------|
| 1 | 0.0 | 1.028 | 206.97 | |
| 10 | 2e-4 | 0.193 | 97.36 | End of 1st cosine cycle |
| 20 | 2e-4 | 0.160 | 81.49 | End of 2nd cycle |
| 30 | 2e-4 | 0.133 | 63.67 | End of 3rd cycle |
| 40 | 2e-4 | 0.132 | 58.41 | End of 4th cycle |
| 45 | 0.0 | 0.109 | 52.36 | |
| 50 | 2e-4 | 0.129 | 72.19 | Cycle peak oscillation |
| 54 | 2e-4 | 0.119 | 51.93 | |
| **58** | **2e-4** | **0.117** | **50.64** | **← Best** |
| 60 | 2e-4 | 0.132 | 59.98 | End of 6th cycle |
| 65 | 0.0 | 0.120 | 51.01 | Final epoch (timeout) |

### Experiment Details
- **Command:** `cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_TIMEOUT_MINUTES=180 SENPAI_MAX_EPOCHS=9999 python train.py --dataset tandemfoil --optimizer lion --lr 2e-4 --cosine-t-max 10 --no-use-ema --model-slices 64 --model-layers 5 --model-hidden-dim 256 --model-heads 4 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --wandb-group tandem-5L256d-lr2e4 --wandb-name "violet/tandem-5L256d-lr2e4-180min" --agent violet`
- **W&B run:** thfga56r
- **Epochs:** 65 (180-min timeout, ~2.77 min/epoch)
- **Peak memory:** ~77.5 GB VRAM (GPU 0)

### What Happened

**The 5L/256d + lr=2e-4 combination did NOT beat the 3L/192d baseline (50.64 vs 45.07), but it did improve over the previous 5L/256d result at lr=3e-4 (52.81 → 50.64).**

Key observations:
1. **lr=2e-4 helped the 5L/256d architecture** — improved from 52.81 (PR #2595, lr=3e-4, T_max=10) to 50.64, confirming that the lower LR provides more stable optimization.
2. **The gains did NOT compound** — the hypothesis that combining 5L/256d + lr=2e-4 would beat 3L/192d + lr=2e-4 (45.07) was wrong. The deeper architecture underperformed by ~12%.
3. **Large oscillation persists** — val_primary ranged from 50-80 even in late training (epochs 50-65), suggesting the 5L/256d model has high sensitivity to the cosine LR restart. The 3L/192d likely converges more smoothly.
4. **Still improving at timeout** — new bests were set at epochs 54 and 58 (late in training), and loss was still decreasing. More epochs *might* help, but the rate of improvement was slowing (~1 point per 10 epochs in the 50s range).
5. **geom_camber_cruise split excels** — at 33.36 (val) / 48.92 (test), this split performed much better than others, suggesting the deeper model captures cruise geometry patterns well but struggles with the harder distribution shifts.

### Analysis

The 3L/192d architecture at lr=2e-4 (45.07) remains superior to 5L/256d at lr=2e-4 (50.64). This suggests:
- **The 5L/256d model may be too deep** for TandemFoil's dataset size, causing optimization difficulties despite the lower LR.
- **3L/192d hits a better capacity-efficiency tradeoff** — enough capacity to learn the pressure field without the optimization challenges of deeper networks.
- The persistent oscillation (50-80 range in late epochs) indicates the cosine restart schedule with T_max=10 may be too aggressive for the deeper model.

### Suggested Follow-ups
1. **Try 4L/224d** — an intermediate depth that might get the best of both worlds.
2. **5L/256d with T_max=20 or 30** — a longer cosine cycle might reduce oscillation and let the deeper model converge better (note: PR #2724 may already test this).
3. **5L/256d with lr=1e-4** — even lower LR might tame the oscillation for the deeper architecture.
4. **5L/256d with gradient clipping** — could help stabilize the deeper model's optimization.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2723 because ADVISOR: Closing — 5L/256d+lr=2e-4 at 50.64 vs baseline 44.72 (now updated). 12.4% worse despite 65 epochs. Key insight: 5L/256d is too deep for TandemFoil at the current dataset scale — optimization is harder and the capacity does not compensate. 3L/192d hits the better capacity-efficiency tradeoff. Violet's insight about geom_camber_cruise excelling (33.36 vs 48.92 test) suggests the deeper model may have distribution-specific strengths, but overall performance is worse. Kakashi #2775 will confirm whether a 5L/256d+T_max=10 variant makes any difference..
